### ISSUE TITLE
🎯 Feature Enhancement: Implement "Sniper Mode" Logic (High Precision & Strict Risk)

### ISSUE THREAD

**Issue by SefaGH** [SefaGH/bearish-alpha-bot#416]
## 📋 Overview
Transition the bot from "Test Mode" (high volume, loose validation) to **"Sniper Mode"**.
This update tightens RSI thresholds, enforces stricter Risk/Reward ratios, and implements probability-based filtering for the PPO Agent to ensure only high-confidence trades are executed.

## 🎯 Objectives
1.  **Filters:** Prevent trades in choppy/neutral markets (RSI 35-65).
2.  **Risk:** Raise R/R requirement from 1.5 to **2.0**.
3.  **AI:** Require **75%+ confidence** from the RL Agent, not just a binary "Buy" signal.
4.  **Safety:** Restrict "Extreme Bypass" to crash scenarios only (RSI < 12).

---

#### A. `src/strategies/adaptive_ob.py`
**Goal:** Lower the entry threshold significantly to catch deep dips only.

*Replace `get_adaptive_rsi_threshold` and `calculate_dynamic_position_size` with:*

```python
    def get_adaptive_rsi_threshold(self, market_regime: Dict) -> float:
        """
        SNIPER MODE: Lower RSI thresholds.
        Waits for deep oversold conditions to avoid catching falling knives.
        """
        # --- SNIPER CHANGE 1: Lower Base ---
        # Old: 45 | New: 32
        base_rsi = float(self.base_cfg.get('adaptive_rsi_base', 32.0))
        
        # Old: 10 | New: 8
        adapt_range = float(self.base_cfg.get('adaptive_rsi_range', 8.0))
        
        trend = market_regime.get('trend', 'neutral')
        momentum = market_regime.get('momentum', 'sideways')
        
        threshold = base_rsi
        
        if trend == 'bullish':
            if momentum == 'strong':
                threshold = base_rsi + 2.0
            else:
                threshold = base_rsi
        elif trend == 'bearish':
            # Be very selective in downtrends
            if momentum == 'strong':
                threshold = base_rsi - 5.0 # Target ~27
            else:
                threshold = base_rsi - 2.0 # Target ~30
        
        # --- SNIPER CHANGE 2: Tighten Ceiling ---
        # Never consider >40 as oversold
        min_threshold = max(20, base_rsi - adapt_range)
        max_threshold = min(40, base_rsi + adapt_range) 
        
        return max(min_threshold, min(max_threshold, threshold))

    def calculate_dynamic_position_size(self, volatility_regime: str, 
                                       base_multiplier: float = 1.0) -> float:
        # --- SNIPER CHANGE 3: Reduce risk in high vol ---
        if volatility_regime == 'high':
            return base_multiplier * 0.50 # Old: 0.75
        elif volatility_regime == 'low':
            return base_multiplier * 1.25
        else:
            return base_multiplier
```

#### B. `src/strategies/adaptive_str.py`
**Goal:** Raise the sell threshold to target peaks/pumps only.

*Replace `get_adaptive_rsi_threshold` with:*

```python
    def get_adaptive_rsi_threshold(self, market_regime: Dict) -> float:
        """
        SNIPER MODE: Higher RSI thresholds.
        Only signals SELL when price is truly overextended.
        """
        # --- SNIPER CHANGE 1: Raise Base ---
        # Old: 50 | New: 68
        base_rsi = float(self.base_cfg.get('adaptive_rsi_base', 68.0))
        
        # Old: 10 | New: 8
        adapt_range = float(self.base_cfg.get('adaptive_rsi_range', 8.0))
        
        trend = market_regime.get('trend', 'neutral')
        momentum = market_regime.get('momentum', 'sideways')
        
        threshold = base_rsi
        
        if trend == 'bearish':
            if momentum == 'strong':
                threshold = base_rsi - 3.0 
            else:
                threshold = base_rsi
        elif trend == 'bullish':
            # Avoid shorting uptrends unless extreme
            if momentum == 'strong':
                threshold = base_rsi + 7.0 # Target ~75
            else:
                threshold = base_rsi + 3.0 # Target ~71
        
        # --- SNIPER CHANGE 2: Tighten Floor ---
        # Never consider <60 as overbought
        min_threshold = max(60, base_rsi - adapt_range)
        max_threshold = min(90, base_rsi + adapt_range)
        
        return max(min_threshold, min(max_threshold, threshold))
```

### 2. Configuration (`config/live_trading_config.yaml`)

**Goal:** Enforce stricter filters globally.

```yaml
risk:
  rr_dynamic:
    enabled: true
    # SNIPER MODE: Aim for 1:2 Risk/Reward
    base_target_rr: 2.0              # Old: 1.5
    lower_bound_rr: 1.2              # Old: 0.8
    upper_bound_rr: 3.0              # Old: 2.0
    
    weights:
      ml_confidence: 0.35            # Increased trust in ML
      rl_agreement: 0.35             # Increased trust in RL
      regime_clarity: 0.2
      volume_strength: 0.05
      momentum_strength: 0.05

ml:
  signal_scoring:
    enabled: true
    min_score_to_trade: 75           # Old: 60 (Require higher confidence)

signals:
  bypass:
    enabled: true
    rsi_oversold_threshold: 12       # Old: 20 (Only bypass on crash)
    rsi_overbought_threshold: 88     # Old: 80 (Only bypass on pump)
  
  oversold_bounce:
    ignore_regime: false             # Enforce regime check
    min_rr_ratio: 1.5                # Old: 0.5
    adaptive_rsi_base: 32
    
  short_the_rip:
    ignore_regime: false             # Enforce regime check
    min_rr_ratio: 1.5                # Old: 0.5
    adaptive_rsi_base: 68
```

### 3. Risk Manager (`src/core/risk_manager.py`)

**Goal:** Disable strategies that perform poorly (<40% win rate).

*Update `_create_default_rules`:*

```python
            # --- SNIPER MODE UPDATE ---
            StrategyPerformanceRule(
                min_win_rate=0.40,  # Increased from 0.35
                performance_monitor=self.performance_monitor
            )
```

### 4. RL Adapter (`src/ml/adapters/ppo_trading_adapter.py`)

**Goal:** Implement "Confidence Threshold". If PPO says "Buy" but isn't sure, treat it as "Hold".

*Replace `get_long_score` with:*

```python
    async def get_long_score(
        self,
        symbol: str,
        *,
        position_fraction: Optional[float] = None,
        normalized_pv: Optional[float] = None,
    ) -> Tuple[float, Dict[str, Any]]:
        """
        Return PPO score for long/flat decision (+ metadata).
        SNIPER MODE: Checks prediction probability/confidence.
        """
        symbol_norm = self._normalize_symbol(symbol)
        
        if not self.cfg.enabled:
            return self.cfg.fallback_score, {"reason": "disabled", "symbol": symbol}
        if not self._is_symbol_supported(symbol_norm):
            return self.cfg.fallback_score, {"reason": "unsupported_symbol", "symbol": symbol}

        await self._ensure_model_loaded()
        if not self._model:
            return self.cfg.fallback_score, {"reason": self._load_error or "model_unavailable"}

        state = await self._build_state(
            symbol_norm,
            position_fraction=position_fraction,
            normalized_pv=normalized_pv,
        )
        if state is None:
            return self.cfg.fallback_score, {"reason": "missing_state"}

        try:
            # --- SNIPER MODE: Probability Based Decision ---
            # Get standard action
            action, _ = self._model.predict(state[np.newaxis, :], deterministic=True)
            action_int = int(action)
            
            # Calculate confidence
            confidence = 1.0 
            if hasattr(self._model, 'policy'):
                import torch as th
                with th.no_grad():
                    obs_tensor = th.as_tensor(state[np.newaxis, :], device=self._model.device)
                    distribution = self._model.policy.get_distribution(obs_tensor)
                    probs = distribution.distribution.probs.cpu().numpy()[0]
                    confidence = float(probs[action_int])

            # --- Sniper Filter ---
            CONFIDENCE_THRESHOLD = 0.75 
            
            final_score = 0.0
            decision = "FLAT"
            
            if action_int == 1:  # Model says BUY
                if confidence >= CONFIDENCE_THRESHOLD:
                    final_score = 1.0  # Strong Buy
                    decision = "LONG"
                else:
                    final_score = 0.5  # Weak Buy -> Ignore
                    decision = "WEAK_LONG_IGNORED"
            else:
                final_score = 0.0 
                decision = "FLAT"

            metadata = {
                "symbol": symbol_norm,
                "action": decision,
                "raw_action": action_int,
                "confidence": confidence,
                "state_tail": {
                    "position_fraction": float(state[-2]),
                    "normalized_pv": float(state[-1]),
                },
                "normalized_symbol": symbol_norm,
                "requested_symbol": symbol,
            }
            
            if action_int == 1 and decision == "WEAK_LONG_IGNORED":
                logger.info(f"🎯 [SNIPER] Ignored weak PPO signal for {symbol}. Conf: {confidence:.2f} < {CONFIDENCE_THRESHOLD}")

            return final_score, metadata

        except Exception as exc:
            logger.warning("PPO prediction failed for %s: %s", symbol_norm, exc)
            return self.cfg.fallback_score, {"reason": "prediction_error", "error": str(exc)}
```

---

## ✅ Acceptance Criteria
- [ ] Bot **refuses** to trade in neutral zones (e.g., RSI 45-55).
- [ ] Bot **refuses** trades where R/R ratio is < 1.2.
- [ ] PPO Adapter logs "WEAK_LONG_IGNORED" for low confidence signals.
- [ ] `live_trading.log` shows significantly reduced signal frequency but higher signal quality logic.
